teps/0091-verified-remote-resources.md

```diff
@@ -0,0 +1,550 @@
+third-party contributors creating Tasks. But those Tasks use builder images
+(external binaries) and the Pipeline Author wants to trust their outputs.
+The Pipeline Author could utilize only verifiable Task Bundles and fail the
+build completely on verification failure, limiting the blast radius of a
+possibly compromised component.
+
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+Task verification using Cosign requires fetching from a possibly external
+OCI repository (the same repository as the builder image), so when a Remote
+Resource is configured to be verified, some overhead is incurred on the order
+of approximately 5 seconds per Remote Resource. Remote Resources that are not
+configured to be verified will have no change in performance.
+
+Tekton Chains will only need to make simple decisions over only locally
+available data so there will be no change to performance.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+TaskRun's `taskRef.resource` will be updated with the following **optional**
+configuration `verification`:
+
+```yaml
+
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+
+spec:
+  taskRef:
+    resolver: bundle
+    resource:
+      image_url: gcr.io/kaniko-project/kaniko-task
+      name: kaniko-build-task
+      verification:
+        signer: [cosign]
+        key: [some-public-key]
+        onError: [ stopAndFail | continue ]
+```
+
+`signer: cosign`. (Required) Additional verification providers can be added in
+the future. Initially, this TEP will support only `cosign`.
+
+`key: [some-public-key]`. (Required) Direct input to Cosign. Cosign supports
+many types of inputs including plain text and various KMS systems.
+
+`onError: [ stopAndFail | continue ]`. (Optional, default `stopAndFail`)
+`stopAndFail` will cause the build to fail if image verification fails, in the
+spirit of secure by default. `continue` will allow the build to continue even
+if verification fails.
+
+TEP-0060 introduces a new ResourceRequest reconsiler that will actually pull
+the remote image and expand it to a Task. This is where the verification
```

Comment:
:+1: thanks for adding the clarifying detail about the location of the verification logic.

I'm going to add an additional Risk to TEP-0060 somewhat related to this: since `ResourceRequest` objects are just CRDs they suffer the same problems as Tekton's other CRDs - they can be modified pretty much any time by someone with the right permissions. So conceivably a Task that's verified by a Resolver might not be the same Task that a TaskRun reconciler receives, since someone with access to the cluster and write-access to ResourceRequests could modify the content moments after resolution + verification completes. It's a different threat model than the one this TEP is addressing but I want to document that Risk somewhere.